internal/rbd/controllerserver.go

```diff
@@ -192,53 +192,49 @@ func (cs *ControllerServer) CreateVolume(ctx context.Context, req *csi.CreateVol
 		return nil, status.Error(codes.Internal, err.Error())
 	}
 
+	rbdSnap, err := cs.checkSnapshotSource(ctx, req, cr)
+	if err != nil {
+		return nil, err
+	}
+
 	found, err := rbdVol.Exists(ctx)
 	if err != nil {
 		if _, ok := err.(ErrVolNameConflict); ok {
 			return nil, status.Error(codes.AlreadyExists, err.Error())
 		}
-
 		return nil, status.Error(codes.Internal, err.Error())
 	}
 	if found {
+		if rbdSnap != nil {
+			err = checkFlatten(ctx, rbdVol, cr)
+			if err != nil {
+				return nil, err
+			}
+		}
 		return buildCreateVolumeResponse(ctx, req, rbdVol)
 	}
 
-	rbdSnap, err := cs.checkSnapshotSource(ctx, req, cr)
-	if err != nil {
-		return nil, err
-	}
-
 	err = reserveVol(ctx, rbdVol, rbdSnap, cr)
 	if err != nil {
 		return nil, status.Error(codes.Internal, err.Error())
 	}
 	defer func() {
 		if err != nil {
-			errDefer := undoVolReservation(ctx, rbdVol, cr)
-			if errDefer != nil {
-				klog.Warningf(util.Log(ctx, "failed undoing reservation of volume: %s (%s)"), req.GetName(), errDefer)
+			if _, ok := err.(ErrFlattenInProgress); !ok {
+				errDefer := undoVolReservation(ctx, rbdVol, cr)
+				if errDefer != nil {
+					klog.Warningf(util.Log(ctx, "failed undoing reservation of volume: %s (%s)"), req.GetName(), errDefer)
+				}
 			}
 		}
 	}()
 
-	err = createBackingImage(ctx, cr, rbdVol, rbdSnap)
+	err = cs.createBackingImage(ctx, cr, rbdVol, rbdSnap)
 	if err != nil {
-		return nil, err
-	}
-
-	if rbdVol.Encrypted {
-		err = rbdVol.ensureEncryptionMetadataSet(rbdImageRequiresEncryption)
-		if err != nil {
-			klog.Errorf(util.Log(ctx, "failed to save encryption status, deleting image %s: %s"),
-				rbdVol, err)
-			if deleteErr := deleteImage(ctx, rbdVol, cr); deleteErr != nil {
-				klog.Errorf(util.Log(ctx, "failed to delete rbd image: %s with error: %v"),
-					rbdVol, deleteErr)
-				return nil, deleteErr
-			}
-			return nil, err
+		if _, ok := err.(ErrFlattenInProgress); ok {
```

Comment:
use `error.Is(err, ErrFlattenInProgress)` as described at https://github.com/golang/go/wiki/ErrorValueFAQ#how-should-i-change-my-error-handling-code-to-work-with-the-new-features